File: EndGate/tests/EndGate.Core.Tests/Utilities/LooperFacts.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using EndGate.Core.Utilities;
using Xunit;

namespace EndGate.Core.Tests
{
    public class LooperFacts
    {
        [Fact]
        public void LooperSingleCallbackWorks()
        {
            long updates = 0;
            var resetEvent = new ManualResetEvent(false);
            var gameLoop = new Looper();
            var looperCallback = new LooperCallback(30, () =>
            {
                updates++;
                if (updates >= 30)
                {
                    resetEvent.Set();
                }
            });
            gameLoop.AddCallback(looperCallback);
            gameLoop.Start();

            Assert.True(resetEvent.WaitOne(TimeSpan.FromSeconds(1.01)));
            gameLoop.Dispose();

            Assert.Equal(updates, 30);
        }

        [Fact]
        public void LooperMultipleCallbacksWork()
        {
            long updates = 0;
            var resetEvent = new ManualResetEvent(false);
            var gameLoop = new Looper();
            var looperCallback1 = new LooperCallback(30, () =>
            {
                updates++;
            });
            var looperCallback2 = new LooperCallback(60, () =>
            {
                updates--;

                if (updates <= -30)
                {
                    resetEvent.Set();
                }
            }); 
            
            gameLoop.AddCallback(looperCallback1);
            gameLoop.AddCallback(looperCallback2);
            gameLoop.Start();

            Assert.True(resetEvent.WaitOne(TimeSpan.FromSeconds(1.01)));
            gameLoop.Dispose();

            // Since looperCallback2 is executing twice as fast as looperCallback1 it will negate the value of updates
            Assert.Equal(updates, -30);
        }
    }
}

File: EndGate/tests/EndGate.Core.Client.JS.Tests/Tests/UnitTests/Utilities/LooperFacts.js
(function (window, lib) {

    QUnit.module("Looper Facts");

    QUnit.asyncTimeoutTest("Looper single callback works.", testUtilities.defaultTestTimeout, function (end, assert, testName) {
        var updates = 0,
            triggered = false,
            gameLoop = new lib.Looper(),
            onComplete = function () {
                assert.equal(updates, 30, "Updates have hit 30!");
                end();
            },
            looperCallback = new lib.LooperCallback(30, function () {
                updates++;
                if (updates >= 30) {
                    triggered = true;
                    gameLoop.Dispose();
                    window.setTimeout(onComplete, 1000);
                }
            });

        gameLoop.AddCallback(looperCallback);
        gameLoop.Start();

        window.setTimeout(function () {
            if (!triggered) {
                assert.ok(false, "Loop did not finish in time!");
                end();
            }
        }, 1050);

        return function () {
            gameLoop.Dispose();
        };
    });

    QUnit.asyncTimeoutTest("Looper multiple callback works.", testUtilities.defaultTestTimeout, function (end, assert, testName) {
        var updates = 0,
            triggered = false,
            gameLoop = new lib.Looper(),
            onComplete = function () {
                assert.equal(updates, -30, "Updates have hit -30!");
                end();
            },
            looperCallback1 = new lib.LooperCallback(30, function () {
                updates++;                
            }),
            looperCallback2 = new lib.LooperCallback(60, function () {
                updates--;

                if (updates <= -30) {
                    triggered = true;
                    gameLoop.Dispose();
                    window.setTimeout(onComplete, 1000);
                }
            });

        gameLoop.AddCallback(looperCallback1);
        gameLoop.AddCallback(looperCallback2);
        gameLoop.Start();

        window.setTimeout(function () {
            if (!triggered) {
                assert.ok(false, "Loop did not finish in time!");
                end();
            }
        }, 1050);

        return function () {
            gameLoop.Dispose();
        };
    });

})(window, EndGate.Core.Utilities);